scipy/optimize/_trustregion_iterative.py

```diff
@@ -0,0 +1,470 @@
+        wp = (1-p[k])/c.T[k, k]
+        wm = (-1-p[k])/c.T[k, k]
+        pp = p[k+1:]+c.T[k+1:, k]*wp
+        pm = p[k+1:]+c.T[k+1:, k]*wm
+
+        if abs(wp)+norm(pp, 1) >= abs(wm)+norm(pm, 1):
+            w[k] = wp
+            p[k+1:] = pp
+        else:
+            w[k] = wm
+            p[k+1:] = pm
+
+    # The system `C v = w` is solved using backward substitution.
+    v = solve_triangular(c, w)
+
+    # Compute vector norms
+    vnorm = norm(v)
+    wnorm = norm(w)
+
+    # Smallest singular value
+    smin = wnorm/vnorm
+
+    # Associated vector
+    zmin = v/vnorm
+
+    return smin, zmin
+
+
+def singular_leading_submatrix(a, c, k):
+    """
+    Compute term that makes the leading ``k`` by ``k``
+    submatrix from `a` singular.
+
+    Parameters
+    ----------
+    a : ndarray
+        Symmetric matrix that is not positive definite.
+    c : ndarray
+        Upper triangular matrix resulting of an incomplete
+        Cholesky decomposition of matrix `a`.
+    k : int
+        Positive integer such that the leading k by k submatrix from
+        `a` is the first non-positive definite leading submatrix.
+
+    Returns
+    -------
+    delta : float
+        Amout that should be added to the element (k, k) of the
+        leading k by k submatrix of ``a`` to make it singular.
+    v : ndarray
+        A vector such that ``v.T B v = 0``. Where B is the matrix A after
+        ``delta`` is added to its element (k, k).
+    """
+
+    # Compute delta
+    delta = np.sum(c[:k-1, k-1]**2) - a[k-1, k-1]
+
+    # Get dimension
+    n = len(a)
+
+    # Inicialize v
+    v = np.zeros(n)
+    v[k-1] = 1
+
+    # Compute the remaining values of v by solving a triangular system
+    if k != 1:
+        v[:k-1] = solve_triangular(c[:k-1, :k-1], -c[:k-1, k-1])
+
+    return delta, v
+
+
+class IterativeSubproblem(BaseQuadraticSubproblem):
+    """Quadratic subproblem solved by the iterative method.
+
+    Notes
+    -----
+    This subproblem solver was based on _[1], _[2] and _[3],
+    which implement similar algorithms. The algorithm is basically
+    that of _[1] but ideas from _[2] and _[3] were also used.
+
+    References
+    ----------
+    .. [1] Conn, A. R., Gould, N. I., & Toint, P. L. (2000).
+           "Trust region methods". Siam. pp. 169-200.
+    .. [2] Nocedal, J., and  Wright, S. (2006).
+           "Numerical optimization". Springer Science &
+           Business Media. pp. 83-91.
+    .. [3] Moré, J. J., & Sorensen, D. C. (1983).
+           Computing a trust region step. SIAM Journal on Scientific
+           and Statistical Computing, 4(3), 553-572.
+    """
+
+    UPDATE_COEF = 0.01
+    EPS = np.finfo(float).eps
+
+    def __init__(self, x, fun, jac, hess=None, hessp=None,
+                 optimization_factor=0.8, trust_region_factor=1.1):
+
+        self.lb = None
+        self.previous_trust_radius = -1
+        self.niter = 0
+        self.easycase_stop_criteria = min(trust_region_factor-1,
```

Comment:
I see. But it looks like you need to use `max` in the first formula. In order to determine `k_hard` we need `(1 - k_easy)^2 <= k_opt` (otherwise `k_hard` goes out from the formula) or `k_easy >= 1 - k_opt**0.5`. So it means that `max` should be used. Can you check that.